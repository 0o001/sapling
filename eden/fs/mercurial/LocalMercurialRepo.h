/*
 *  Copyright (c) 2016, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
#pragma once
#include <folly/String.h>

namespace facebook {
namespace eden {

// Represents what we know about a mercurial repo on the local disk.
// We may add more information in here in the future.
// The thought is that we'll keep this repo checked out on the null
// rev, so that it is roughly equivalent to a bare checkout.
// See also LocalMercurialRepoAndRev
class LocalMercurialRepo {
  // The path to the directory that contains the .hg dir for the repo
  folly::fbstring localPath_;

 public:
  explicit LocalMercurialRepo(folly::StringPiece path);

  const folly::fbstring& getPath() const;
};
}
}
