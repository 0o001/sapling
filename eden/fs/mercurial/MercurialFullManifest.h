/*
 *  Copyright (c) 2016, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
#pragma once
#include "eden/utils/LeaseCache.h"
#include "eden/utils/PathFuncs.h"

namespace facebook {
namespace eden {

class LocalMercurialRepoAndRev;

// Maintains information about the monolithic hg manifest.
// During construction we use `hg files` to discover the directories.
// Later, we fill out basic stat(2) like information on demand.
// This doesn't perform too well with the hg of today, but is closer
// to the access pattern that we looking for.  I fully expect that we'll
// tear this up as we iterate.
class MercurialFullManifest {
 public:
  // For a directory within the manifest, the list of files and child dirs.
  // Both are sorted
  struct DirListing {
    folly::fbvector<folly::fbstring> files;
    folly::fbvector<folly::fbstring> dirs;
  };
  // For a file, the basic info we can use to fill out a `struct stat`
  struct FileInfo {
    mode_t mode;
    size_t size;

    FileInfo(mode_t mode, size_t size);
  };
  static std::unique_ptr<MercurialFullManifest> parseManifest(
      LocalMercurialRepoAndRev& repo);

  // Returns information about a given file.
  // This is backed by the LeaseCache
  folly::Future<std::shared_ptr<FileInfo>> getFileInfo(RelativePathPiece path);

  // Returns an object containing the list of entries for a given dir
  const DirListing& getListing(const folly::fbstring& path);

  // An optimization that can bulk load the FileInfo for a given dir
  folly::Future<folly::Unit> prefetchFileInfoForDir(RelativePathPiece path);

  // Obtain the contents of the specified path.
  // For symlinks this is the target of the symlink.
  // For plain files this is the content of the file itself.
  folly::Future<std::string> catFile(RelativePathPiece path);

 private:
  explicit MercurialFullManifest(LocalMercurialRepoAndRev& repo);
  LocalMercurialRepoAndRev& repo_;
  std::unordered_map<folly::fbstring, DirListing> dirs_;
  LeaseCache<folly::fbstring, FileInfo> fileInfo_;

  void load();
  DirListing& getOrMakeEntry(folly::StringPiece path);
  folly::Future<std::shared_ptr<FileInfo>> fetchFileInfo(
      const folly::fbstring& name);
};
}
}
