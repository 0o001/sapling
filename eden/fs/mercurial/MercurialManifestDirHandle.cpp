/*
 *  Copyright (c) 2016, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
#include "eden/fs/mercurial/MercurialManifestDirHandle.h"
#include "eden/fuse/Inodes.h"
#include "eden/utils/PathFuncs.h"

namespace facebook {
namespace eden {

using InodeNameManager = fusell::InodeNameManager;

MercurialManifestDirHandle::MercurialManifestDirHandle(
    fuse_ino_t parent,
    fuse_ino_t ino,
    std::shared_ptr<LocalMercurialRepoAndRev> repo,
    RelativePathPiece path)
    : parent_(parent), ino_(ino), repo_(repo), path_(path) {}

folly::Future<fusell::DirList> MercurialManifestDirHandle::readdir(
    fusell::DirList&& list, off_t off) {
  const auto& listing = repo_->getManifest().getListing(path_.value());
  struct stat st;

  memset(&st, 0, sizeof(st));

  // The first two slots are linkage to parent and self
  constexpr ino_t SELF_ENTRY = 0;
  constexpr ino_t PARENT_ENTRY = 1;
  constexpr ino_t FIRST_FILE_ENTRY = 2;

  if (off == SELF_ENTRY) {
    st.st_ino = ino_;
    if (!list.add(".", st, ++off)) {
      return std::move(list);
    }
  }

  if (off == PARENT_ENTRY) {
    st.st_ino = parent_;
    if (!list.add("..", st, ++off)) {
      return std::move(list);
    }
  }

  // The next range is for the files in this dir
  const auto& files = listing.files;
  // We don't yet know enough to mark this as S_IFLNK or S_IFREG
  st.st_mode = 0;
  while (off - FIRST_FILE_ENTRY < files.size()) {
    const auto& name = files[off - FIRST_FILE_ENTRY];
    auto node =
        InodeNameManager::get()->getNodeByName(ino_, PathComponentPiece(name));
    st.st_ino = node->getNodeId();

    if (!list.add(name.c_str(), st, ++off)) {
      return std::move(list);
    }
  }

  // And the final range is for the dirs
  const auto& dirs = listing.dirs;
  ino_t first_dir_entry = FIRST_FILE_ENTRY + files.size();
  constexpr size_t MAX_PREFETCH = 6;
  size_t n_prefetched = 0;

  // We know enough to flag these as dirs.  Only the type matters, not
  // the permissions because the syscall is only mapping these to
  // DT_XXX (https://www.daemon-systems.org/man/DTTOIF.3.html)
  st.st_mode = S_IFDIR;
  while (off - first_dir_entry < dirs.size()) {
    const auto& name = dirs[off - first_dir_entry];
    PathComponentPiece namepiece(name);
    auto node = InodeNameManager::get()->getNodeByName(ino_, namepiece);
    st.st_ino = node->getNodeId();

    if (n_prefetched++ < MAX_PREFETCH) {
      auto full_name = path_ + namepiece;
      repo_->getManifest().prefetchFileInfoForDir(full_name);
    }
    if (!list.add(name.c_str(), st, ++off)) {
      return std::move(list);
    }
  }
  return std::move(list);
}

folly::Future<fusell::Dispatcher::Attr> MercurialManifestDirHandle::setattr(
    const struct stat&, int) {
  folly::throwSystemErrorExplicit(EROFS,
                                  "cannot set attributes in the manifest");
}

folly::Future<folly::Unit> MercurialManifestDirHandle::fsyncdir(bool) {
  // Nothing to do; this is a read only handle
  return folly::Unit{};
}

folly::Future<fusell::Dispatcher::Attr> MercurialManifestDirHandle::getattr() {
  fusell::Dispatcher::Attr attr;
  attr.st.st_mode = S_IFDIR | 0755;
  attr.st.st_ino = ino_;
  return attr;
}
}
}
