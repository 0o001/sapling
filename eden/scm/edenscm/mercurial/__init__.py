# Portions Copyright (c) Facebook, Inc. and its affiliates.
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2.

# __init__.py - Startup and module loading logic for Mercurial.
#
# Copyright 2015 Gregory Szorc <gregory.szorc@gmail.com>
#
# This software may be used and distributed according to the terms of the
# GNU General Public License version 2 or any later version.

from __future__ import absolute_import

import os
import sys

# Allow 'from mercurial import demandimport' to keep working.
from edenscm import hgdemandimport


# pyre-fixme[11]: Annotation `hgdemandimport` is not defined as a type.
demandimport = hgdemandimport

__all__ = []

if getattr(sys, "platform") == "win32":
    configdir = os.path.join(
        getattr(os, "environ").get("PROGRAMDATA") or r"\ProgramData",
        "Facebook",
        "Mercurial",
    )
else:
    configdir = "/etc/mercurial"


def shoulduselegacy(name):
    legacy = getattr(os, "environ").get("HGLEGACY")
    if legacy is not None:
        return name in legacy.split()
    else:
        return os.path.lexists(os.path.join(configdir, "legacy.%s" % name))
