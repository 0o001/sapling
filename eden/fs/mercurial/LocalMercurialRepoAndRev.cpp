/*
 *  Copyright (c) 2016, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
#include "LocalMercurialRepoAndRev.h"
#include "MercurialFullManifest.h"

namespace facebook {
namespace eden {

LocalMercurialRepoAndRev::LocalMercurialRepoAndRev(
    folly::StringPiece rev, std::shared_ptr<LocalMercurialRepo> repo)
    : rev_(rev.fbstr()),
      repo_(repo),
      manifest_(MercurialFullManifest::parseManifest(*this)) {}

const folly::fbstring& LocalMercurialRepoAndRev::getRev() const { return rev_; }

std::shared_ptr<LocalMercurialRepo> LocalMercurialRepoAndRev::getRepo() const {
  return repo_;
}

MercurialFullManifest& LocalMercurialRepoAndRev::getManifest() {
  return *manifest_.get();
}
}
}
