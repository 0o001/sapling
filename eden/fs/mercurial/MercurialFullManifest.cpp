/*
 *  Copyright (c) 2016, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
#include "MercurialFullManifest.h"
#include "LocalMercurialRepoAndRev.h"
#include "eden/utils/PathFuncs.h"
#include "eden/utils/SortedInsert.h"
#include <folly/Subprocess.h>
#include <wangle/concurrent/GlobalExecutor.h>

DEFINE_int32(hg_manifest_file_cache_size,
             65536,
             "maximum number of file entries to cache");

namespace facebook {
namespace eden {

MercurialFullManifest::MercurialFullManifest(LocalMercurialRepoAndRev& repo)
    : repo_(repo),
      fileInfo_(
          FLAGS_hg_manifest_file_cache_size,
          [=](const folly::fbstring& name) { return fetchFileInfo(name); }) {
  load();
}

MercurialFullManifest::DirListing& MercurialFullManifest::getOrMakeEntry(
    folly::StringPiece name) {
  auto key = name.fbstr();

  auto find = dirs_.find(key);
  if (find == dirs_.end()) {
    // Recursively build out parents if missing
    auto parent_dir = dirname(name);
    if (parent_dir != name) {
      auto& parent = getOrMakeEntry(dirname(name));
      // Add ourselves to the parent
      sorted_insert(parent.dirs, basename(name).str(), CompareString());
    }
  }
  return dirs_[key];
}

std::unique_ptr<MercurialFullManifest> MercurialFullManifest::parseManifest(
    LocalMercurialRepoAndRev& repo) {
  std::unique_ptr<MercurialFullManifest> manifest(
      new MercurialFullManifest(repo));
  manifest->load();
  return manifest;
}

void MercurialFullManifest::load() {
  auto path = repo_.getRepo()->getPath();
  LOG(INFO) << "Parsing file list for " << path << " @ " << repo_.getRev();
  size_t num_files = 0;

  folly::Subprocess proc({"hg", "files", "-r", repo_.getRev().toStdString()},
                         folly::Subprocess::pipeStdout()
                             .chdir(path.toStdString())
                             .closeOtherFds()
                             .usePath());

  auto read_cb = folly::Subprocess::readLinesCallback(
      [this, &num_files, &path](int fd, folly::StringPiece line) {
        if (fd == STDOUT_FILENO) {
          if (line.empty()) {
            return false;
          }
          line.removeSuffix("\n");
          folly::StringPiece dir = dirname(line);
          folly::StringPiece filename = basename(line);

          // This will create the dir node on demand
          auto& entry = getOrMakeEntry(dir);
          // and add this file to its list
          sorted_insert(entry.files, filename.fbstr(), CompareString());
          num_files++;
        } else {
          LOG(ERROR) << "[" << path << "] hg files -r " << repo_.getRev()
                     << " stderr: " << line;
        }
        return false; // Keep reading from the child
      });
  proc.communicate(std::ref(read_cb), [](int, int) {
    // Don't write to the child
    return true;
  });
  proc.waitChecked();
  LOG(INFO) << "build tree with " << dirs_.size() << " dirs";
}

MercurialFullManifest::FileInfo::FileInfo(mode_t mode, size_t size)
    : mode(mode), size(size) {}

static mode_t flags_to_mode(folly::StringPiece flags) {
  if (flags.find('d') != std::string::npos) {
    return S_IFDIR | 0755;
  }

  mode_t mode = S_IFREG;

  if (flags.find('l') != std::string::npos) {
    mode = S_IFLNK;
  }

  if (flags.find('x') != std::string::npos) {
    mode |= 0755;
  } else {
    mode |= 0644;
  }

  return mode;
}

folly::Future<std::shared_ptr<MercurialFullManifest::FileInfo>>
MercurialFullManifest::getFileInfo(RelativePathPiece name) {
  return fileInfo_.get(name.copy().value());
}

folly::Future<folly::Unit> MercurialFullManifest::prefetchFileInfoForDir(
    RelativePathPiece name) {
  return via(wangle::getCPUExecutor().get()).then([ =, name = name.copy() ] {
    const auto& listing = getListing(name.value());
    if (listing.files.empty()) {
      return;
    }

    std::vector<std::string> args = {"hg",
                                     "files",
                                     "-r",
                                     repo_.getRev().toStdString(),
                                     "-vT",
                                     "{size}\\0{flags}\\0{abspath}\\n"};

    size_t n = 0;
    for (auto& file : listing.files) {
      auto full_name = (name + PathComponentPiece(file)).value().toStdString();
      if (fileInfo_.exists(full_name)) {
        continue;
      }
      args.emplace_back(std::move(full_name));
      n++;
    }

    if (n == 0) {
      // Everything is already warm enough
      return;
    }

    LOG(INFO) << "Running hg files on dir '" << name << "' for " << n
              << " files";
    folly::Subprocess proc(args,
                           folly::Subprocess::pipeStdout()
                               .chdir(repo_.getRepo()->getPath().toStdString())
                               .closeOtherFds()
                               .usePath());

    auto read_cb = folly::Subprocess::readLinesCallback([this](
        int fd, folly::StringPiece line) {
      if (fd == STDOUT_FILENO) {
        if (line.empty()) {
          return false;
        }

        line.removeSuffix("\n");
        folly::fbvector<folly::StringPiece> fields;
        folly::split('\0', line, fields);

        if (fields.size() != 3) {
          throw std::runtime_error(
              folly::to<std::string>("bad output from hg files: ", line));
        }

        fileInfo_.set(fields[2].fbstr(),
                      std::make_shared<FileInfo>(flags_to_mode(fields[1]),
                                                 folly::to<size_t>(fields[0])));
      }
      return false; // Keep reading from the child
    });
    proc.communicate(std::ref(read_cb), [](int, int) {
      // Don't write to the child
      return true;
    });
    proc.wait();
  });
}

folly::Future<std::shared_ptr<MercurialFullManifest::FileInfo>>
MercurialFullManifest::fetchFileInfo(const folly::fbstring& name) {
  // First, if it is a dir then we can very quickly return its info
  if (dirs_.find(name) != dirs_.end()) {
    return std::make_shared<FileInfo>(S_IFDIR | 0755, 0);
  }

  return via(wangle::getCPUExecutor().get()).then([=] {
    std::vector<std::string> args = {"hg",
                                     "files",
                                     "-r",
                                     repo_.getRev().toStdString(),
                                     "-vT",
                                     "{size}\\0{flags}\\n",
                                     name.toStdString()};
    LOG(INFO) << "Running hg files on " << name;
    folly::Subprocess proc(args,
                           folly::Subprocess::pipeStdout()
                               .chdir(repo_.getRepo()->getPath().toStdString())
                               .closeOtherFds()
                               .usePath());
    auto output = proc.communicate();
    proc.waitChecked();

    folly::StringPiece line(output.first);

    line.removeSuffix("\n");
    folly::fbvector<folly::StringPiece> fields;
    folly::split('\0', line, fields);

    if (fields.size() != 2) {
      throw std::runtime_error(
          folly::to<std::string>("bad output from hg files: ", line));
    }

    return std::make_shared<FileInfo>(flags_to_mode(fields[1]),
                                      folly::to<size_t>(fields[0]));

  });
}

const MercurialFullManifest::DirListing& MercurialFullManifest::getListing(
    const folly::fbstring& name) {
  return dirs_.at(name);
}

folly::Future<std::string> MercurialFullManifest::catFile(
    RelativePathPiece path) {
  return via(wangle::getCPUExecutor().get()).then([
    =,
    path = path.stringPiece().str()
  ] {
    folly::Subprocess proc(
        {"hg", "cat", "-r", repo_.getRev().toStdString(), path},
        folly::Subprocess::pipeStdout()
            .pipeStderr()
            .chdir(repo_.getRepo()->getPath().toStdString())
            .closeOtherFds()
            .usePath());
    auto output = proc.communicate();
    try {
      proc.waitChecked();
      if (!output.second.empty()) {
        LOG(ERROR) << "stderr not empty while running `hg cat -r "
                   << repo_.getRev() << " " << path << "`: " << output.second;
      }
      return output.first;
    } catch (const std::exception& e) {
      LOG(ERROR) << "Exception while running `hg cat -r " << repo_.getRev()
                 << " " << path << "`: " << e.what() << ", " << output.second;
      throw; // Will typically bubble up as EIO to the consumer
    }
  });
}
}
}
