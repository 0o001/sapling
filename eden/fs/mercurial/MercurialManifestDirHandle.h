/*
 *  Copyright (c) 2016, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
#pragma once
#include "LocalMercurialRepoAndRev.h"
#include "eden/fuse/DirHandle.h"

namespace facebook {
namespace eden {

// Represents an opendir()'d handle to a dir in the hg manifest
class MercurialManifestDirHandle : public fusell::DirHandle {
  fuse_ino_t parent_;
  fuse_ino_t ino_;
  std::shared_ptr<LocalMercurialRepoAndRev> repo_;
  RelativePath path_;

 public:
  MercurialManifestDirHandle(
      fuse_ino_t parent,
      fuse_ino_t ino,
      std::shared_ptr<LocalMercurialRepoAndRev> repo,
      RelativePathPiece path);
  folly::Future<fusell::DirList> readdir(fusell::DirList&& list,
                                         off_t off) override;

  folly::Future<fusell::Dispatcher::Attr> setattr(const struct stat& attr,
                                                  int to_set) override;
  folly::Future<folly::Unit> fsyncdir(bool datasync) override;
  folly::Future<fusell::Dispatcher::Attr> getattr() override;
};
}
}
