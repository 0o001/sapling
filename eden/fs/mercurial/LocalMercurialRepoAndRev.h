/*
 *  Copyright (c) 2016, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the BSD-style license found in the
 *  LICENSE file in the root directory of this source tree. An additional grant
 *  of patent rights can be found in the PATENTS file in the same directory.
 *
 */
#pragma once
#include "LocalMercurialRepo.h"
#include "MercurialFullManifest.h"

namespace facebook {
namespace eden {

class MercurialFullManifest;

// References a repo at a specific revision.
// Since we don't intend to have a full checkout maintained at any revision
// and we may also end up serving multiple users and checkouts, we need
// a way to reference the source of the history as well as a current
// revision.
// In addition to referencing the repo and revision, this provides
// an accessor to the manifest.
// At present we only have access to the full manifest information in
// mercurial, and we materialize this during construction.  The intention
// is that we'll do this all lazily in the future when hg has support
// for querying it in that fashion.
class LocalMercurialRepoAndRev
    : public std::enable_shared_from_this<LocalMercurialRepoAndRev> {
  folly::fbstring rev_;
  std::shared_ptr<LocalMercurialRepo> repo_;
  std::unique_ptr<MercurialFullManifest> manifest_;

 public:
  LocalMercurialRepoAndRev(folly::StringPiece rev,
                           std::shared_ptr<LocalMercurialRepo> repo);
  const folly::fbstring& getRev() const;
  std::shared_ptr<LocalMercurialRepo> getRepo() const;

  MercurialFullManifest& getManifest();
};
}
}
