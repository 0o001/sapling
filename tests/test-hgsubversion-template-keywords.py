# no-check-code -- see T24862348

import test_hgsubversion_util
from mercurial import commands, error, ui


try:
    from mercurial import templatekw

    templatekw.keywords
except ImportError:
    templatekw = None

try:
    from mercurial import revset

    revset.methods
except ImportError:
    revset = None


class CapturingUI(ui.ui):
    def __init__(self, *args, **kwds):
        super(CapturingUI, self).__init__(*args, **kwds)
        self._output = ""

    def write(self, msg, *args, **kwds):
        self._output += msg


class TestLogKeywords(test_hgsubversion_util.TestBase):
    @test_hgsubversion_util.requiresmodule(templatekw)
    def test_svn_keywords(self):
        defaults = {"date": None, "rev": None, "user": None, "graph": True}
        repo = self._load_fixture_and_fetch("two_revs.svndump")

        # we want one commit that isn't from Subversion
        self.commitchanges([("foo", "foo", "frobnicate\n")])

        ui = CapturingUI()
        commands.log(ui, repo, template=("  rev: {rev} svnrev:{svnrev}\n"), **defaults)
        print(ui._output)
        self.assertEqual(
            ui._output.strip(),
            """
  rev: 2 svnrev:
@
|
  rev: 1 svnrev:3
o
|
  rev: 0 svnrev:2
o
""".strip(),
        )

    @test_hgsubversion_util.requiresmodule(revset)
    @test_hgsubversion_util.requiresmodule(templatekw)
    def test_svn_revsets(self):
        repo = self._load_fixture_and_fetch("two_revs.svndump")

        # we want one commit that isn't from Subversion
        self.commitchanges([("foo", "foo", "frobnicate\n")])

        defaults = {"date": None, "rev": ["fromsvn()"], "user": None}

        ui = CapturingUI()
        commands.log(ui, repo, template="{rev}:{svnrev} ", **defaults)
        self.assertEqual(ui._output, "0:2 1:3 ")

        defaults = {"date": None, "rev": ["svnrev(2)"], "user": None}

        ui = CapturingUI()
        commands.log(ui, repo, template="{rev}:{svnrev} ", **defaults)
        self.assertEqual(ui._output, "0:2 ")

        defaults = {"date": None, "rev": ["fromsvn(1)"], "user": None}

        self.assertRaises(
            error.ParseError,
            commands.log,
            self.ui(),
            repo,
            template="{rev}:{svnrev} ",
            **defaults
        )

        defaults = {"date": None, "rev": ["svnrev(1, 2)"], "user": None}

        self.assertRaises(
            error.ParseError,
            commands.log,
            self.ui(),
            repo,
            template="{rev}:{svnrev} ",
            **defaults
        )


if __name__ == "__main__":
    import silenttestrunner

    silenttestrunner.main(__name__)
