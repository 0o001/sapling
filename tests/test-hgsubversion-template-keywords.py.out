  rev: 2 svnrev:
@
|
  rev: 1 svnrev:3
o
|
  rev: 0 svnrev:2
o


